### PR TITLE
vm health checking

### PR DESCRIPTION
**Pull request by adiprerepa**
VM Health Checking Final Implementation.

Design Doc: https://docs.google.com/document/d/1G4s_PrmvkwCKoH_kXVeSUAszsaj3ZfpIlraSdeTSPKU/

cc @howardjohn @stevenctl 

agent implementation: #27273

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -280,6 +281,20 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream discovery.AggregatedD
...
+	// to health info
+	if request.TypeUrl == health.HealthInfoTypeURL {
+		event := HealthEvent{}
+		if request.ErrorDetail == nil {
+			event.Healthy = true
+		} else {
+			event.Healthy = false
+			event.Message = request.ErrorDetail.Message
+		}
+		s.InternalGen.UpdateWorkloadEntryHealth(con.proxy, event)
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:294]: Strongly suggest moving it out to `processRequest`. And make it running in a new goroutine to unblock the normal xds process

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -33,6 +33,7 @@ import (
 	"istio.io/istio/pilot/pkg/model"
 	"istio.io/istio/pilot/pkg/networking/util"
 	v3 "istio.io/istio/pilot/pkg/xds/v3"
+	"istio.io/istio/pkg/istio-agent/health"
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:36]: It is bad, we should never let pilot/istiod import from agent. 

Move the healthcheck type to pilot/pkg/xds/v3/model.go

File: pilot/pkg/xds/workloadentry.go [hunk 3]
```diff
@@ -121,6 +130,36 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
+// UpdateWorkloadEntryHealth updates the associated WorkloadEntries health status
+// based on the corresponding health check performed by istio-agent.
+func (sg *InternalGen) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthEvent) {
+	if !features.WorkloadEntryHealthChecks {
+		return
+	}
+	// we assume that the workload entry exists
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		return
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/workloadentry.go:142]: maybe add some debug info

File: pilot/pkg/xds/workloadentry.go [hunk 4]
```diff
@@ -121,6 +130,36 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
+	}
+
+	// get previous status
+	cfg := sg.Store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
+	if cfg == nil {
+		return
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/workloadentry.go:148]: same here

File: pilot/pkg/xds/ads.go (hunk 2 above)
Comment (adiprerepa) [reply pilot/pkg/xds/ads.go:36]: Should I import the TypeUrl in Istio-agent then?

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (adiprerepa) [reply pilot/pkg/xds/ads.go:294]: Why do we want to move it out of `processRequest`? Where else would we move it? This seems like an appropriate place.

@howardjohn thoughts?

File: pilot/pkg/xds/workloadentry.go [hunk 5]
```diff
@@ -121,6 +130,38 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
 }
 
+// UpdateWorkloadEntryHealth updates the associated WorkloadEntries health status
+// based on the corresponding health check performed by istio-agent.
+func (sg *InternalGen) UpdateWorkloadEntryHealth(proxy *model.Proxy, event HealthEvent) {
+	if !features.WorkloadEntryHealthChecks {
+		return
+	}
+	// we assume that the workload entry exists
+	entryName := autoregisteredWorkloadEntryName(proxy)
```
Comment (stevenctl) [inline pilot/pkg/xds/workloadentry.go:140]: Is the manually created case going to be handled in a different PR?

Comment (adiprerepa) [reply pilot/pkg/xds/workloadentry.go:140]: i was under the impression that all of them would be auto-registered...How do I get the name of the manually created WorkloadEntry?

Comment (stevenctl) [reply pilot/pkg/xds/workloadentry.go:140]: https://docs.google.com/document/d/1G4s_PrmvkwCKoH_kXVeSUAszsaj3ZfpIlraSdeTSPKU/edit#bookmark=id.t6ju2m1mphe6

Something can be added to the proxyConfig, we can use an ISTIO_META_, or we could lookup by ip. For the last option, I'm not sure what the perf is on field selectors like this; I doubt they're indexed so it could be a bit slow at scale.

File: pilot/pkg/xds/workloadentry.go [hunk 6]
```diff
@@ -121,6 +130,38 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
+
+	// get previous status
+	cfg := sg.Store.Get(gvk.WorkloadEntry, entryName, proxy.Metadata.Namespace)
+	if cfg == nil {
+		adsLog.Debugf("cfg nil when getting WorkloadEntry %v", entryName)
+		return
+	}
+
+	// replace the updated status
+	status := cfg.Status.(*v1alpha1.IstioStatus)
```
Comment (stevenctl) [inline pilot/pkg/xds/workloadentry.go:154]: I think I saw issues when modifying something returned directly from Store. You may need to DeepCopy (which probably doesn't include Status at the moment)

File: pilot/pkg/xds/workloadentry.go [hunk 7]
```diff
@@ -20,9 +20,11 @@ import (
 	"strings"
 	"time"
 
+	"github.com/gogo/protobuf/types"
```
Comment (stevenctl) [inline pilot/pkg/xds/workloadentry.go:23]: Can you add some unit test coverage?

File: pilot/pkg/xds/workloadentry.go (hunk 5 above)
Comment (adiprerepa) [reply pilot/pkg/xds/workloadentry.go:140]: sorry, for the third option, lookup to where? Is there a DNS thing for WorkloadEntry names?

Comment (howardjohn) [reply pilot/pkg/xds/workloadentry.go:142]: The workload entries have IP address in the spec

Comment (howardjohn) [reply pilot/pkg/xds/workloadentry.go:142]: we can do auto registered ones only for now to make this simpler

File: pilot/pkg/xds/workloadentry.go [hunk 8]
```diff
@@ -121,6 +130,49 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
+	if !features.WorkloadEntryHealthChecks {
+		return
+	}
+	// we assume that the workload entry exists
+	// if auto registration does not exist, try looking
+	// up in NodeMetadata
+	entryName := autoregisteredWorkloadEntryName(proxy)
+	if entryName == "" {
+		// todo make this env var a constant somewhere
+		name, ok := proxy.Metadata.ProxyConfig.ProxyMetadata["WORKLOAD_ENTRY_NAME"]
```
Comment (howardjohn) [inline pilot/pkg/xds/workloadentry.go:145]: The value should be prefixed with ISTIO_META_. Then it will be put in node metadata, and you can access by `proxy.metadata.WorkloadEntryName` (once you add the field)

File: pilot/pkg/xds/workloadentry.go [hunk 9]
```diff
@@ -121,6 +130,49 @@ func (sg *InternalGen) QueueUnregisterWorkload(proxy *model.Proxy) {
...
+		if !ok {
+			// at this point we could not find the WorkloadEntry Name,
+			// just return.
+			adsLog.Errorf("unable to derive WorkloadEntry Name, blank name")
```
Comment (howardjohn) [inline pilot/pkg/xds/workloadentry.go:149]:
```suggestion
			adsLog.Errorf("unable to derive WorkloadEntry for health update for %v", proxy.ID)
```

Comment (howardjohn) [reply pilot/pkg/xds/workloadentry.go:149]: We should make sure all logs have enough context about what it was doing and what connection it was

File: pkg/config/model.go [hunk 10]
```diff
@@ -261,6 +261,43 @@ func DeepCopy(s Spec) Spec {
 
 type Status interface{}
 
+// is this the same as spec?
```
Comment (howardjohn) [inline pkg/config/model.go:264]: Yes, they are both just `interface{}`, can we make it a single DeepCopy method?

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/ads.go:293]: I mean move it directly to `processRequest`. This should not be part of `shouldRespond`.

File: pilot/pkg/xds/workloadentry.go (hunk 8 above)
Comment (adiprerepa) [reply pilot/pkg/xds/workloadentry.go]: if we are doing non-auto registration cases later this will be in a different pr

**Review by howardjohn — APPROVED** [review]
Can you move the logic out of shouldRespond into processRequest? We can add a hook like `continue := preprocessRequest(con, req)` there, so that we do not need have so much type-specific bloat in the mainline code

**Comment by adiprerepa** [comment]
/retest

for the goddamn millionth time

File: pilot/pkg/xds/workloadentry_test.go [hunk 11]
```diff
@@ -90,9 +92,7 @@ func TestNonAutoregisteredWorkloads(t *testing.T) {
 }
 
 func TestAutoregistrationLifecycle(t *testing.T) {
-	if features.WorkloadEntryCleanupGracePeriod != 200*time.Millisecond {
-		features.WorkloadEntryCleanupGracePeriod = 200 * time.Millisecond
-	}
+	features.WorkloadEntryCleanupGracePeriod = 200 * time.Millisecond
```
Comment (stevenctl) [inline pilot/pkg/xds/workloadentry_test.go:95]: This was there to prevent a weird race I saw when running the test with `count`
